Redefine Dust and add a discard_rate

morcos opened the pull request:
The definition of dust is redefined to remove the factor of 3.

Dust is redefined to be the value of an output such that it would
cost that value in fees to (create and) spend the output at the dust
relay rate.  The previous definition was that it would cost 1/3 of the
value.  The default dust relay rate is correspondingly increased to
3000 sat/kB so the actual default dust output value of 546 satoshis
for a non-segwit output remains unchanged.  This commit is a refactor
only unless a dustrelayfee is passed on the commandline in which case
that number now needs to be increased by a factor of 3 to get the same
behavior.  -dustrelayfee is a hidden command line option.

Note: It's not exactly a refactor due to edge case changes in rounding
as evidenced by the required change to the unit test.

A discard_rate is added which defaults to 10,000 sat/kB

Any change output which would be dust at the discard_rate you are
willing to discard completely and add to fee (as well as continuing to
pay the fee that would have been needed for creating the change)

This would be a nice addition for 0.15 and I think will remain useful for 0.16 with the new coin selection algorithms in discussion, but its not crucial.

It does add translation strings, but we could (should?) avoid that by hiding the option

morcos committed:
Add a discard_rate

Any change output which would be dust at the discard_rate you are
willing to discard completely and add to fee (as well as continuing to
pay the fee that would have been needed for creating the change).

gmaxwell reviewed (approved):
ACK

morcos committed:
Remove factor of 3 from definition of dust.

This redefines dust to be the value of an output such that it would
cost that value in fees to (create and) spend the output at the dust
relay rate.  The previous definition was that it would cost 1/3 of the
value.  The default dust relay rate is correspondingly increased to
3000 sat/kB so the actual default dust output value of 546 satoshis
for a non-segwit output remains unchanged.  This commit is a refactor
only unless a dustrelayfee is passed on the commandline in which case
that number now needs to be increased by a factor of 3 to get the same
behavior.  -dustrelayfee is a hidden command line option.

Note: It's not exactly a refactor due to edge case changes in rounding
as evidenced by the required change to the unit test.

morcos commented:
This PR has been changed to redefine dust and move the discard rate slightly lower.  See new PR description.

morcos commented:
After discussion with @gmaxwell, just wanted to point out that the edge case rounding issues would only be with non-standard dust rates.  Since the default dust rate was 1000 sat/kB and the division in question was by 1000, then there are no rounding issues regardless of the output in question.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/10817/commits/28da790417d0a4d544988e12d5ce9033b0b97f87

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -2501,6 +2503,15 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
+    unsigned int highest_target = estimator.HighestTargetTracked(FeeEstimateHorizon::LONG_HALFLIFE);
+    CFeeRate discard_rate = estimator.estimateSmartFee(highest_target, nullptr /* FeeCalculation */, false /* conservative */);
+    discard_rate = std::min(discard_rate, CWallet::m_discard_rate);
+    discard_rate = std::max(discard_rate, ::dustRelayFee);
```
In commit "Add a discard_rate"

Maybe it should be an init error to specify `-dustrelayfee` greater than `-discardrate`, because the `-discardrate` value is ignored in this case.

morcos replied:
I just informed in the option help rather than erroring..

ryanofsky reviewed (commented):
utACK 28da790417d0a4d544988e12d5ce9033b0b97f87. Much less confusing without the factor of 3.

gmaxwell reviewed (approved):
re-ACK

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -2501,6 +2503,15 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
+static CFeeRate GetDiscardRate(const CBlockPolicyEstimator& estimator)
+{
+    unsigned int highest_target = estimator.HighestTargetTracked(FeeEstimateHorizon::LONG_HALFLIFE);
+    CFeeRate discard_rate = estimator.estimateSmartFee(highest_target, nullptr /* FeeCalculation */, false /* conservative */);
```
Dont you need to check that this wasnt an error return (ie 0)?

instagibbs replied:
i.e., failure to get a real estimate means we'll go to dustRelayFee instead of static discard rate

morcos replied:
Yes I suppose that would be a slight improvement.  Stupid 0 as error value

TheBlueMatt reviewed (commented):
This is awesome, though it would be nice to document the limits for discardfee (must be at least dust relay fee, will not go higher than estimate fee).

morcos commented:
Addressed comments

([discard.ver2](https://github.com/morcos/bitcoin/commits/discard.ver2)) -> f4d00e6 ([discard.ver2.squash](https://github.com/morcos/bitcoin/commits/discard.ver2.squash))

TheBlueMatt commented:
utACK f4d00e63f7951469c840edede3f675d07249f62a

achow101 commented:
utACK f4d00e63f7951469c840edede3f675d07249f62a

gmaxwell reviewed (approved):
re-ACK

ryanofsky reviewed (commented):
utACK f4d00e63f7951469c840edede3f675d07249f62a. Has new safeguard against failing fee estimate & more comments and documentation.

instagibbs commented:
utACK with new safeguard logic
